### ISSUE TITLE
Installation problem

### ISSUE THREAD

**Issue by khayatlab** [scipion-em/scipion-installer#19]
Hi,

I'm getting the following error when trying to install:

python3 -m pip install scipion-installer
Downloading/unpacking scipion-installer
  Could not find a version that satisfies the requirement scipion-installer (from versions: 0a0, 0.1a0, 0.2a0, 0.3a0, 0.4a0, 0.5a0, 0.6a0, 0.7a0, 0.9a0, 0.10a0, 0.10.1a0, 0.10.2a0, 0.10.3a0, 0.10.4a0, 0.10.5a0, 0.10.6a0, 0.10.7a0, 0.10.8a0, 1.0.0b0, 1.0.1b0, 1.0.2b0, 1.0.3b0, 1.0.4b0, 1.0.5b0, 1.0.6b0, 1.0.6b1, 1.0.6b2, 1.0.6b3, 1.0.6b4, 1.0.6b5, 1.0.6b6)
Cleaning up...
No distributions matching the version for scipion-installer
Storing debug log for failure in /home/rkhayat/.pip/pip.log

Any suggestions? Thanks.

Reza

**Comment by pconesa** [comment]
Weird. . What happens if you specify the version:

python3 -m pip install scipion-installer==1.0.6b6

?

**Comment by khayatlab** [comment]
Hi,
This seems to work but requires a root install. Is it possible to install scipion without root permission? I've tried:

python3 -m pip install --install-option="--prefix=/home/rkhayat/Applications/scipion-installer" scipion-installer==1.0.6b6

Files are downloaded, but the script ends with an error:
    Installing installscipion script to /home/rkhayat/Applications/scipion-installer/bin
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 283, in run
    requirement_set.install(install_options, global_options, root=options.root_path)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 1436, in install
    requirement.install(install_options, global_options, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 672, in install
    self.move_wheel_files(self.source_dir, root=root)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 902, in move_wheel_files
    pycompile=self.pycompile,
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 206, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 193, in clobber
    os.makedirs(destsubdir)
  File "/usr/lib/python3.4/os.py", line 237, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.4/dist-packages/virtualenv-20.0.28.dist-info'

Storing debug log for failure in /home/rkhayat/.pip/pip.log

**Comment by azazellochg** [comment]
The easiest way is `pip install --user scipion-installer` or `python3 -m pip install --user scipion-installer`

**Comment by pconesa** [comment]
This page may help.

We recommend to make a conda installation if no root access.

https://scipion-em.github.io/docs/scipion3_instalation/docs/scipion-modes/install-from-sources.html

**Comment by khayatlab** [comment]
Thanks, I've tried this. There is some syntax error when trying to install following Ubuntu with Conda instructions. 

Following Ubuntu with Virtualenv works a little better but fails during XMIPP setup:

 g++ 4.8 detected
 g++ -c -w -mtune=native -march=native -std=c++11 -O3 xmipp_test_main.cpp -o xmipp_test_main.o -I../ -I/home/rkhayat/Applications/scipion3/.scipion3env/include -I/home/rkhayat/anaconda3/include/python3.7m -I/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/numpy/core/include
xmipp_test_main.cpp:3:18: fatal error: hdf5.h: No such file or directory
 #include <hdf5.h>
                  ^
compilation terminated.
 Check the INCDIRFLAGS, CXX, CXXFLAGS and PYTHONINCFLAGS in xmipp.conf
 If some of the libraries headers fail, try installing fftw3_dev, tiff_dev, jpeg_dev, sqlite_dev, hdf5, pthread
 Cannot compile
Possible solutions
In Ubuntu: sudo apt-get -y install libsqlite3-dev libfftw3-dev libhdf5-dev libopencv-dev python3-dev python3-numpy python3-scipy python3-mpi4py
In Manjaro: sudo pacman -Syu install hdf5 python3-numpy python3-scipy --noconfirm
Please, see 'https://scipion-em.github.io/docs/docs/scipion-modes/install-from-sources.html#step-2-dependencies' for more information about libraries dependencies.

Remember to re-run './xmipp config' after installing libraries in order to take into account the new system configuration.
 rm xmipp_test_main*

Check failed! Something wrong with the configuration.

Traceback (most recent call last):
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/__main__.py", line 443, in <module>
    main()
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/__main__.py", line 287, in main
    installPluginMethods()
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/install/install_plugin.py", line 231, in installPluginMethods
    plugin.installBin({'args': ['-j', numberProcessor]})
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/install/plugin_funcs.py", line 166, in installBin
    environment.execute()
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/install/funcs.py", line 748, in execute
    self._executeTargets(targetList)
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/install/funcs.py", line 690, in _executeTargets
    tgt.execute()
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/install/funcs.py", line 220, in execute
    command.execute()
  File "/home/rkhayat/Applications/scipion3/.scipion3env/lib/python3.7/site-packages/scipion/install/funcs.py", line 161, in execute
    "running '%s')" % (t, cmd))
AssertionError: target '/home/rkhayat/Applications/scipion3/software/em/xmipp/bin/xmipp_reconstruct_significant' not built (after running 'cd . && ./xmipp config noAsk &&./xmipp check_config && ./xmipp compileAndInstall N=4 && ln -srf build /home/rkhayat/Applications/scipion3/software/em/xmipp && cd - && touch installation_finished && rm bindings_linked 2> /dev/null')
Error at main: target '/home/rkhayat/Applications/scipion3/software/em/xmipp/bin/xmipp_reconstruct_significant' not built (after running 'cd . && ./xmipp config noAsk &&./xmipp check_config && ./xmipp compileAndInstall N=4 && ln -srf build /home/rkhayat/Applications/scipion3/software/em/xmipp && cd - && touch installation_finished && rm bindings_linked 2> /dev/null')

Something went wrong running:
 cd /home/rkhayat/Applications/scipion3 &&
/home/rkhayat/anaconda3/bin/python -m virtualenv --python=python3 .scipion3env &&
. /home/rkhayat/Applications/scipion3/.scipion3env/bin/activate &&
cd /home/rkhayat/Applications/scipion3 &&
mkdir -p software/lib &&
mkdir -p software/bindings &&
mkdir -p software/em &&
export SCIPION_HOME=/home/rkhayat/Applications/scipion3 &&
pip install scipion-pyworkflow &&
pip install scipion-app &&
python -m scipion installp -p scipion-em-xmipp -j 4
Installation cancelled.

**Comment by azazellochg** [comment]
Hi,

as suggested in the error message, you are missing hdf5 library headers.
You need to install  libhdf5-dev.
Alternatively, if you do not have root, have a look at conda installation
as suggested above.

Best regards,
Grigory

--------------------------------------------------------------------------------
Grigory Sharov, Ph.D.

MRC Laboratory of Molecular Biology,
Francis Crick Avenue,
Cambridge Biomedical Campus,
Cambridge CB2 0QH, UK.
tel. +44 (0) 1223 267228 <+44%201223%20267228>
e-mail: gsharov@mrc-lmb.cam.ac.uk

**Comment by pconesa** [comment]
@khayatlab , the log you are posting is for a virtualenv installation. It requires to install (apt-get install) some libraries in case they are missing like libhdf5-dev.

If you cannot install those dependencies:

`sudo apt-get install gcc g++ make libopenmpi-dev python3-tk libfftw3-dev libhdf5-dev libtiff-dev libjpeg-dev libsqlite3-dev openjdk-8-jdk`

I'd suggest what @azazellochg is suggesting.

Maybe you can post the "syntax error when trying to install following Ubuntu with Conda instructions"?

**Comment by khayatlab** [comment]
Thank you both for the help. It would be best for me to pick one installation method so as to expedite the installation.  I get the following error when trying the Conda instructions under a tcsh shell:

_python -m scipioninstaller /home/rkhayat/Applications/scipion3 -j 4
conda detected. Favouring it. If you want a virtualenv installation cancel installation and pass --venv .
This is an early version of the installer. So far only works for developers installing an unstable version.Are you sure you want to continue? (y/n): y
sh: 1: eval: setenv: not found
sh: 2: eval: setenv: not found
sh: 3: eval: setenv: not found
sh: 4: eval: setenv: not found
sh: 34: eval: Syntax error: word unexpected (expecting ")")
Something went wrong running:
 eval "$(/home/rkhayat/anaconda3/bin/conda shell.tcsh hook)" &&
conda create  -n .scipion3env python=3 &&
conda activate .scipion3env &&
cd /home/rkhayat/Applications/scipion3 &&
mkdir -p software/lib &&
mkdir -p software/bindings &&
mkdir -p software/em &&
export SCIPION_HOME=/home/rkhayat/Applications/scipion3 &&
pip install scipion-pyworkflow &&
pip install scipion-app &&
python -m scipion installp -p scipion-em-xmipp -j 4
Installation cancelled._

I receive the following error when trying to install using a bash shell:
_python -m scipioninstaller /home/rkhayat/Applications/scipion3 -j 4
conda detected. Favouring it. If you want a virtualenv installation cancel installation and pass --venv .
This is an early version of the installer. So far only works for developers installing an unstable version.Are you sure you want to continue? (y/n): y
sh: 1: eval: setenv: not found
sh: 2: eval: setenv: not found
sh: 3: eval: setenv: not found
sh: 4: eval: setenv: not found
sh: 34: eval: Syntax error: word unexpected (expecting ")")
Something went wrong running:
 eval "$(/home/rkhayat/anaconda3/bin/conda shell.tcsh hook)" &&
conda create  -n .scipion3env python=3 &&
conda activate .scipion3env &&
cd /home/rkhayat/Applications/scipion3 &&
mkdir -p software/lib &&
mkdir -p software/bindings &&
mkdir -p software/em &&
export SCIPION_HOME=/home/rkhayat/Applications/scipion3 &&
pip install scipion-pyworkflow &&
pip install scipion-app &&
python -m scipion installp -p scipion-em-xmipp -j 4
Installation cancelled._

**Comment by pconesa** [comment]
Thanks @khayatlab I also think this is related to way we are trying to activate conda, which might not be needed if conda is available in the path.

**Comment by khayatlab** [comment]
Thanks. How do I address this?

**Comment by azazellochg** [comment]
@pconesa just pushed a fix, try `pip install --user -U scipion-installer` to update the installer and run it again.

**Comment by khayatlab** [comment]
Thanks. There is now another problem. 
0. conda activate (using bash)
1. export CXX_CUDA=g++-8
2. export PATH=$PATH:/usr/local/cuda-8.0/bin
3. pip install --user -U scipion-installer
4. export PATH=$PATH:/home/rkhayat/.local/bin
5. python -m scipioninstaller /home/rkhayat/Applications/scipion3 -j 4

Many programs install and following error received:
._#"
. # To activate this environment, use"
.#
.#     $ conda activate .scipion3env
.#
.# To deactivate an active environment, use
.#
.#     $ conda deactivate


CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

.    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.

IMPORTANT: You may need to close and restart your shell after running 'conda init'.


Something went wrong running:
 conda create  -n .scipion3env python=3 &&
conda activate .scipion3env &&
cd /home/rkhayat/Applications/scipion3 &&
mkdir -p software/lib &&
mkdir -p software/bindings &&
mkdir -p software/em &&
export SCIPION_HOME=/home/rkhayat/Applications/scipion3 &&
pip install scipion-pyworkflow &&
pip install scipion-app &&
python -m scipion installp -p scipion-em-xmipp -j 4
Installation cancelled._

**Comment by pconesa** [comment]
I've uploaded a new version of the installer:
pip install .....WTF!! @azazellochg you are another Sauron! (impressive).

As @azazellochg wrote, I've released another version of the installer, avoiding the "eval" part".

Still there is an "eval command" later, that I expect to fail but only when launching Scipion and installation should have finished.

Here I'm not sure to avoid the "eval" command too ...thus conda has to be active and initialized always as a requirement....or keep it and see why it didn't work for you the eval, @khayatlab .

So, in short, @khayatlab:

1.- Update the installer
2.- Try conda installation again
3.- If it works, try to launch scipion3:
     3.a: My guess is that it will fail the same way, so edit scipion3 file and change:
        ` 'eval "$(/extra/miniconda3/condabin/conda shell.bash hook)" && conda activate .scipion3env && ' `--> `'conda activate .scipion3env && '`

**Comment by pconesa** [comment]
I see! conda doesn't get initialized properly unless we run an initialized shell....I guess that is why we had eval.

@khayatlab , what is the output of echo $SHELL?

**Comment by pconesa** [comment]
And which OS are you in?

**Comment by khayatlab** [comment]
???? Its /bin/tcsh

**Comment by khayatlab** [comment]
Not sure why it's not switching to bash. 
OS version is:
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.2 LTS
Release:        14.04
Codename:       trusty

**Comment by pconesa** [comment]
in tcsh, does setenv exists?

**Comment by khayatlab** [comment]
Yes.

**Comment by pconesa** [comment]
do you have conda initialized in your .bashrc or tcsh equivalent?

**Comment by khayatlab** [comment]
I'm trying both, and both give me the same error I posted last. 
This is the content of my .tcshrc file

.# >>> conda initialize >>>
.# !! Contents within this block are managed by 'conda init' !!
if ( -f "/home/rkhayat/anaconda3/etc/profile.d/conda.csh" ) then
    source "/home/rkhayat/anaconda3/etc/profile.d/conda.csh"
else
    setenv PATH="/home/rkhayat/anaconda3/bin:$PATH"
endif
.# <<< conda initialize <<<





This is the conda content of my .bashrc file

.# >>> conda initialize >>>
.# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/rkhayat/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/rkhayat/anaconda3/etc/profile.d/conda.sh" ]; then
        . "/home/rkhayat/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/rkhayat/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
.# <<< conda initialize <<<

**Comment by azazellochg** [comment]
@khayatlab could you reopen your shell, run `conda init tsch`, reopen it again then do `conda activate`. Does this give an error?

**Comment by khayatlab** [comment]
I get the following error when I open a new shell
Illegal variable name.

This is because conda init tcsh   changes the content of my .tcshrc file to:

.# >>> conda initialize >>>
.# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/rkhayat/anaconda3/bin/conda' 'shell.tcsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/rkhayat/anaconda3/etc/profile.d/conda.sh" ]; then
        . "/home/rkhayat/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/rkhayat/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
.# <<< conda initialize <<<




conda is no longer recognized

**Comment by pconesa** [comment]
Ok, so there is something wrong there.

In a tcsh shell where "conda info" fails, could try:

eval "$(/home/rkhayat/anaconda3/bin/conda shell.tcsh hook)"
conda info


From here: https://github.com/conda/conda/issues/7980#issuecomment-492784093

**Comment by pconesa** [comment]
I'm removing the latest release since it does not fixes anything.

**Comment by pconesa** [comment]
Ok conda info always work...we need "conda activate something"

**Comment by pconesa** [comment]
eval "$(/extra/miniconda3/bin/conda shell.bash hook)" works for me in a non initialized bash shell.

**Comment by khayatlab** [comment]
I get the following error when I try any of the eval commands

Illegal variable name

**Comment by azazellochg** [comment]
Looks like the conda initialize block has errors in there, like you setup tcsrc but then source sh. What a mess.. 
Shall we try  a normal shell (bash)?

1) remove conda initialize block from tcshrc and bashrc
2) open new console and start bash
3) run `source /home/rkhayat/anaconda3/etc/profile.d/conda.sh`
4) run `conda init`
5) run `conda activate`
6) open new bash console

This should not give errors.

**Comment by pconesa** [comment]
There was a bug in conda initializing tcsh: https://github.com/conda/conda/issues/8584

Which version of conda have you got?

**Comment by khayatlab** [comment]
@azazellochg this takes me back to the initial bash setup where I did not get an error with conda activate, but received an error during installation. I think the problem is that typing echo $SHELL returns /bin/tcsh   I don't think I'm running bash (?)

@pconesa I am running conda 4.6.14

**Comment by khayatlab** [comment]
@pconesa I believe you are correct, it's the conda version 4.6.14
which version of conda should I install?

**Comment by pconesa** [comment]
I think there are 2 issues here:

1.- conda initialization at @khayatlab machine.
2.- scipion installation under tcsh: We are using this kind of command in the installer (eval "$(/home/rkhayat/anaconda3/bin/conda shell.tcsh hook)" and is not compatible with csh shells. As mentioned here: https://stackoverflow.com/a/43752030/3157958

So, if you fix 1, just install update the installer again (I'm releasing a new version like the previous one) it should work in bash shells.

**Comment by pconesa** [comment]
conda usually warns about it:

````
pablo@youyou:~/software/scipion$ conda update

CondaValueError: no package names supplied
# If you want to update to a newer version of Anaconda, type:
#
# $ conda update --prefix /extra/miniconda3 anaconda
 ````

Alternatively latestfrom conda site should work.

**Comment by khayatlab** [comment]
Updated conda to version 4.8.3
conda update  --prefix /home/rkhayat/anaconda3 anaconda 

conda info  (works)
...
          conda version : 4.8.3
...


python -m scipioninstaller /home/rkhayat/Applications/scipion3 -j 4
conda detected. Favouring it. If you want a virtualenv installation cancel installation and pass --venv .
This is an early version of the installer. So far only works for developers installing an unstable version.Are you sure you want to continue? (y/n): y
sh: 1: eval: setenv: not found
sh: 2: eval: setenv: not found
sh: 3: eval: setenv: not found
sh: 4: eval: setenv: not found
sh: 34: eval: Syntax error: word unexpected (expecting ")")
Something went wrong running:
 eval "$(/home/rkhayat/anaconda3/bin/conda shell.tcsh hook)" &&
conda create  -n .scipion3env python=3 &&
conda activate .scipion3env &&
cd /home/rkhayat/Applications/scipion3 &&
mkdir -p software/lib &&
mkdir -p software/bindings &&
mkdir -p software/em &&
export SCIPION_HOME=/home/rkhayat/Applications/scipion3 &&
pip install scipion-pyworkflow &&
pip install scipion-app &&
python -m scipion installp -p scipion-em-xmipp -j 4
Installation cancelled.

**Comment by azazellochg** [comment]
You did not update the installer

**Comment by khayatlab** [comment]
I did update:
pip install --user -U scipion-installer

python -m scipioninstaller /home/rkhayat/Applications/scipion3 -j 4 
conda detected. Favouring it. If you want a virtualenv installation cancel installation and pass --venv .
This is an early version of the installer. So far only works for developers installing an unstable version.Are you sure you want to continue? (y/n): y
sh: 1: eval: setenv: not found
sh: 2: eval: setenv: not found
sh: 3: eval: setenv: not found
sh: 4: eval: setenv: not found
sh: 34: eval: Syntax error: word unexpected (expecting ")")
Something went wrong running:
 eval "$(/home/rkhayat/anaconda3/bin/conda shell.tcsh hook)" &&
conda create  -n .scipion3env python=3 &&
conda activate .scipion3env &&
cd /home/rkhayat/Applications/scipion3 &&
mkdir -p software/lib &&
mkdir -p software/bindings &&
mkdir -p software/em &&
export SCIPION_HOME=/home/rkhayat/Applications/scipion3 &&
pip install scipion-pyworkflow &&
pip install scipion-app &&
python -m scipion installp -p scipion-em-xmipp -j 4
Installation cancelled.

**Comment by azazellochg** [comment]
hm, maybe it didn't work. The eval command was removed. Try pip uninstall
scipion-installer and then pip install --user scipion-installer..
Best regards,
Grigory

--------------------------------------------------------------------------------
Grigory Sharov, Ph.D.

MRC Laboratory of Molecular Biology,
Francis Crick Avenue,
Cambridge Biomedical Campus,
Cambridge CB2 0QH, UK.
tel. +44 (0) 1223 267228 <+44%201223%20267228>
e-mail: gsharov@mrc-lmb.cam.ac.uk

**Comment by khayatlab** [comment]
Same error. I may be downloading an installer that is different than what you are updating. I'm installing:
Successfully installed scipion-installer-1.0.6b8

**Comment by azazellochg** [comment]
ok. that's the latest version.
The last error means you are not running bash / your $SHELL is not pointing
to bash, but tcsh instead. Did you start bash before doing all this?
Best regards,
Grigory

--------------------------------------------------------------------------------
Grigory Sharov, Ph.D.

MRC Laboratory of Molecular Biology,
Francis Crick Avenue,
Cambridge Biomedical Campus,
Cambridge CB2 0QH, UK.
tel. +44 (0) 1223 267228 <+44%201223%20267228>
e-mail: gsharov@mrc-lmb.cam.ac.uk

**Comment by pconesa** [comment]
Pip list will list your packages. There ckech you've got either b6 or b8 version of the installer.

Also, what is the output of echo $SHELL before running the installer?

**Comment by pconesa** [comment]
Ok, version is the latest, should work. The issue now is why the installer falls to tcsh shell.

**Comment by khayatlab** [comment]
1. The default shell is tcsh
2. I type bash in terminal to switch over to temporary bash. The temporary shell can read bash scripts. conda is installed properly (e.g. conda info returns information on conda)
3. Typing echo $SHELL returns /bin/tcsh 
4. What is the file that causes this problem? Is it written in bash?

**Comment by pconesa** [comment]
Maybe I misunderstood the behaviour of $SHELL variable. 

Can you confirm that in your bash, temporary shell, your SHELL variable still says tcsh?

**Comment by pconesa** [comment]
If you pass the installer -dry ...it will print all the commands.

The one with the eval is the one failing on tcsh.

I think you can go one by one and skipping the problematic part (eval ...), keeping the conda activate .scipion3env

Finally we will need to tweak the launcher, probably removing again the eval part might work.

**Comment by pconesa** [comment]
Hi @khayatlab , another option is to set bash as the default shell. Is this something you've intentionally decided (to have tcsh as a defalult shell)?

If you want to go this route this should work

1.- chsh -s /bin/bash
2.- logout + login

**Comment by pconesa** [comment]
How did this go?
